packages/ui/src/composables/useValidation.ts

```diff
@@ -66,12 +69,28 @@ export const useValidation = (
     const rules = flatten(props.rules)
 
     normalizeValidationRules(rules, props.modelValue)
-      .forEach((validationResult: boolean | string) => {
+      .forEach((validationResult: validationResult) => {
         if (isString(validationResult)) {
           errorMessages = [...errorMessages, validationResult]
           error = true
         } else if (validationResult === false) {
           error = true
+        } else if (validationResult instanceof Promise) {
+          innerLoadingState.value = true
+
+          validationResult
+            .then(promisedValidationResult => {
+              if (isString(promisedValidationResult)) {
+                errorMessages = [...errorMessages, promisedValidationResult]
+                error = true
+              }
+            })
+            .catch(e => { throw new Error(e) })
+            .finally(() => {
+              computedErrorMessages.value = [...errorMessages]
+              computedError.value = error
+              innerLoadingState.value = false
+            })
         }
       })
 
```

Comment:
Need to deal with `Promise.all`, looks like !error will be returned before promises resolved.

Reply:
Also, would be nice to have dirty state, to prevent extra validation requests. Remember that we validate a whole form with a lot of fields.